hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientTransactionManagerServiceImpl.java

```diff
@@ -106,23 +108,50 @@ public String getGroupName() {
     }
 
     public ClientConnection connect() throws Exception {
-        Exception lastError = null;
-        int count = 0;
-        while (count < RETRY_COUNT) {
+        long startTimeMillis = System.currentTimeMillis();
+        AbstractClientInvocationService invocationService = (AbstractClientInvocationService) client.getInvocationService();
+        ClientConfig clientConfig = client.getClientConfig();
+        boolean smartRouting = clientConfig.getNetworkConfig().isSmartRouting();
+
+        Exception lastException = null;
+        while (System.currentTimeMillis() - startTimeMillis < invocationService.getInvocationTimeoutMillis()) {
+            if (!client.getLifecycleService().isRunning()) {
+                throw new HazelcastClientNotActiveException("Client is shutdown");
+            }
             try {
-                final Address randomAddress = getRandomAddress();
-                return (ClientConnection) client.getConnectionManager().getOrConnect(randomAddress);
+                if (smartRouting) {
+                    return tryConnectSmart();
+                } else {
+                    return tryConnectUnisocket();
+                }
             } catch (Exception e) {
-                lastError = e;
+                lastException = e;
+                if (e instanceof HazelcastClientOfflineException) {
+                    throw e;
+                }
             }
-            count++;
+            Thread.sleep(invocationService.getInvocationRetryPauseMillis());
         }
-        throw lastError;
+        throw lastException;
     }
 
-    private Address getRandomAddress() {
-        Member member = loadBalancer.next();
-        if (member == null) {
+    private ClientConnection tryConnectUnisocket() throws IOException {
+        ClientConnection connection = client.getConnectionManager().getOwnerConnection();
+
+        if (connection != null) {
+            return connection;
+        }
+        return throwException(false);
+    }
+
+    private ClientConnection throwException(boolean smartRouting) {
+        ClientConfig clientConfig = client.getClientConfig();
+        ClientConnectionStrategyConfig connectionStrategyConfig = clientConfig.getConnectionStrategyConfig();
+        ClientConnectionStrategyConfig.ReconnectMode reconnectMode = connectionStrategyConfig.getReconnectMode();
+        if (reconnectMode.equals(ClientConnectionStrategyConfig.ReconnectMode.ASYNC)) {
```

Comment:
What is the situation for ReconnectMode.ON ? Should it be waiting before throwing the exception?

Reply:
I have a test for reconnect mode on 
https://github.com/hazelcast/hazelcast/pull/11812/files/9c8c3347dc74eee7f4070271eb8bbabdec964ebb#diff-c5d24f27e63ca1c625af589dd927d567R61

We are expecting it to throw exception after invocation timeout.